src/test/disconnected_transactions.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+//
+#include <test/util/setup_common.h>
+#include <kernel/disconnected_transactions.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_FIXTURE_TEST_SUITE(disconnected_transactions, TestChain100Setup)
+
+//! Tests that DisconnectedBlockTransactions limits its own memory properly
+BOOST_AUTO_TEST_CASE(disconnectpool_memory_limits)
+{
+    if (sizeof(void*) != 8) {
+        BOOST_TEST_MESSAGE("Exiting disconnected_transactions memory usage test early due to unsupported arch");
+    }
+
+    // Use the coinbase transactions from TestChain100Setup. It doesn't matter whether these
+    // transactions would realistically be in a block together, they just need distinct txids and
+    // uniform size for this test to work.
+    std::vector<CTransactionRef> block_vtx(m_coinbase_txns);
+    BOOST_CHECK_EQUAL(block_vtx.size(), 100);
+
+    // Roughly estimate sizes to sanity check that DisconnectedBlockTransactions::DynamicMemoryUsage
+    // is within an expected range.
+
+    // Overhead for the hashmap depends on number of buckets
+    std::unordered_map<uint256, CTransaction*, SaltedTxidHasher> temp_map;
+    temp_map.reserve(2);
+    const size_t MAP_2{memusage::DynamicUsage(temp_map)};
+    temp_map.reserve(100);
+    const size_t MAP_100{memusage::DynamicUsage(temp_map)};
+
+    const size_t TX_USAGE{RecursiveDynamicUsage(block_vtx.front())};
+    for (const auto& tx : block_vtx) BOOST_CHECK_EQUAL(RecursiveDynamicUsage(tx), TX_USAGE);
+
+    const size_t ENTRY_USAGE_OVERESTIMATE{TX_USAGE + 16 * memusage::MallocUsage(sizeof(void*))};
+
+    // DisconnectedBlockTransactions that's just big enough for 1 transaction.
+    {
+        DisconnectedBlockTransactions disconnectpool{MAP_2 + ENTRY_USAGE_OVERESTIMATE};
+        // Just 2 transactions because the allocation depends on the number of txns passed in.
+        std::vector<CTransactionRef> two_txns({block_vtx.at(0), block_vtx.at(1)});
+        auto evicted_txns = disconnectpool.AddTransactionsFromBlock(two_txns);
+        BOOST_CHECK(disconnectpool.DynamicMemoryUsage() <= MAP_2 + ENTRY_USAGE_OVERESTIMATE);
+
+        // Only 1 transaction can be kept
```

Comment:
It's not immediately obvious to me why this is expected behaviour, when we instantiated `disconnectpool` to the memory usage of a map of 2 + overhead. At the moment, I think this is the case because (as per my previous comment) the `MAP_x` vars don't include the map nodes, but that seems unintuitive/unexpected?